task_sdk/src/airflow/sdk/execution_time/comms.py

```diff
@@ -0,0 +1,103 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+r"""
+Communication protocol between the Supervisor and the task process
+==================================================================
+
+* All communication is done over stdout/stdin in the form of "JSON lines" (each
+  message is a single JSON document terminated by `\n` character)
+* Messages from the subprocess are all log messages and are sent directly to the log
+* No messages are sent to task process except in response to a request. (This is because the task process will
+  be running user's code, so we can't read from stdin until we enter our code, such as when requesting an XCom
+  value etc.)
```

Comment:
the Supervisor _could_ be written to be async, but the task subprocess itself can't be async right now as task code is sync. (We'll need to do something async around the triggerer)

The Supervisor is a akin to an HTTP proxy for this traffic, but also has other responsibilities (such as watching the process for exting etc.)

Reply:
Okay so then this is a direct proxy-ing - if upload line bandwidth is limited, this would then limit the performance of task execution as all log messages are directly "write though" as stream? Or is there some buffering (would increase performance but potentially lose messages on crash?)

Reply:
Log messages aren't currently handled in this PR -- or at least they aren't being sent upstream to the API server, log messages are just being collected from the subprocess (as json) and then set to stdout of the Supervisor.

"shortly"/in the next PR all those logs will be written to local disk on the worker in files as they do today, and _then_ we can see about improving things for logs to S3 and of running tasks etc.